exercises/parser_combinators/descr.md

```diff
@@ -0,0 +1,636 @@
+interprets it as the corresponding operation on integers.
+
+**Question.** Define a parser combinator `chainl1` of type
+`'a parser -> 'a op parser -> 'a parser` such that the parser
+`chainl1 p op` recognizes a nonempty sequence of `p`'s,
+separated with `op`'s,
+and interprets it as a left-associative chain of applications
+of operators to values.
+
+**Question 3.** Define a parser `sum` of type `int parser` that recognizes and
+evaluates a **sum**, that is, a nonempty sequence of numbers, separated by
+additive operators. For instance, the string `"24-12-12"` should be evaluated
+to the single integer result `0`.
+
+We now wish to introduce the multiplicative operators as well as the ability
+to enclose a subexpression within parentheses. In other words, we would like
+to parse the language described by the following **context-free grammar**:
+
+```
+  expr ::= number
+         | expr additive_op expr
+         | expr multiplicative_op expr
+         | '(' expr ')'
+```
+
+Unfortunately, this grammar cannot directly be used as a guide in the
+definition of a parser. Indeed, it exhibits several flaws. To begin with, it
+is ambiguous: it does not indicate that every operator should be considered
+left-associative, nor does it indicate that multiplicative operators have
+greater priority than additive operators. Furthermore, it is left-recursive:
+several right-hand sides in the definition of `expr` begin with `expr` itself.
+A parser whose structure is modeled directly after this grammar would diverge.
+
+In order to work around these problems, a common solution is to use a slightly
+more complex grammar, where three syntactic categories are distinguished,
+namely **atoms**, **factors**, and **terms**:
+
+```
+  atom   ::= number
+           | '(' term ')'
+  factor ::= atom ( additive_op atom )*
+  term   ::= factor ( multiplicative_op factor )*
+```
+
+This grammar can be transcribed in English as follows:
+
+* An **atom** is either a number or a term enclosed in parentheses.
+
+* A **factor** is a nonempty sequence of atoms, separated with additive operators.
+
+* A **term** is a nonempty sequence of factors, separated with multiplicative operators.
+
+This grammar is still recursive, because `term` depends on `factor` which
+depends on `atom` which depends on `term`. However, it is not left-recursive:
+along this cycle, at least one input symbol (namely, an opening parenthesis)
+must be consumed.
+
+**Question 4.** Define a parser `term` of type `int parser` that recognizes
+and evaluates a **term**, as defined by the above grammar. For instance, the
+string `"49-12*(2+2)"` should be evaluated to the single integer result `1`.
+
+*Hint.* One approach is to define `atom`, `factor` and `term` as three
+mutually recursive functions. An alternative approach is to use the parser
+combinator `fix`, whose type is `('a parser -> 'a parser) -> 'a parser`,
+and whose definition is given to you.
```

Comment:
in the prelude, I guess?

Reply:
At the moment, it is not in the prelude; it is part of the template.